hypothesis-python/src/hypothesis/stateful.py

```diff
@@ -543,12 +584,63 @@ def available(self, data):
         return bool(machine.bundle(self.name))
 
 
-class BundleConsumer(Bundle[Ex]):
-    def __init__(self, bundle: Bundle[Ex]) -> None:
+# class Bundle(SearchStrategy[Ex]):
+#     """A collection of values for use in stateful testing.
+
+#     Bundles are a kind of strategy where values can be added by rules,
+#     and (like any strategy) used as inputs to future rules.
+
+#     The ``name`` argument they are passed is the they are referred to
+#     internally by the state machine; no two bundles may have
+#     the same name. It is idiomatic to use the attribute
+#     being assigned to as the name of the Bundle::
+
+#         class MyStateMachine(RuleBasedStateMachine):
+#             keys = Bundle("keys")
+
+#     Bundles can contain the same value more than once; this becomes
+#     relevant when using :func:`~hypothesis.stateful.consumes` to remove
+#     values again.
+
+#     If the ``consume`` argument is set to True, then all values that are
+#     drawn from this bundle will be consumed (as above) when requested.
+#     """
+
+#     def __init__(self, name: str, *, consume: bool = False) -> None:
+#         self.name = name
+#         self.__reference_strategy = BundleReferenceStrategy(name, consume=consume)
+
+#     def do_draw(self, data):
+#         machine = data.draw(self_strategy)
+#         # print("Bundle do_draw machine: ", machine)
+#         reference = data.draw(self.__reference_strategy)
+#         # print("Bundle do_draw reference: ", reference)
+#         return machine.names_to_values[reference.name]
+
+#     def __repr__(self):
+#         consume = self.__reference_strategy.consume
+#         if consume is False:
+#             return f"Bundle(name={self.name!r})"
+#         return f"Bundle(name={self.name!r}, {consume=})"
+
+#     def calc_is_empty(self, recur):
+#         # We assume that a bundle will grow over time
+#         return False
+
+#     def available(self, data):
+#         # ``self_strategy`` is an instance of the ``st.runner()`` strategy.
+#         # Hence drawing from it only returns the current state machine without
+#         # modifying the underlying buffer.
+#         machine = data.draw(self_strategy)
+#         return bool(machine.bundle(self.name))
+
+
+class BundleConsumer(Bundle):
+    def __init__(self, bundle: Bundle) -> None:
         super().__init__(bundle.name, consume=True)
 
 
-def consumes(bundle: Bundle[Ex]) -> SearchStrategy[Ex]:
+def consumes(bundle: Bundle) -> SearchStrategy:
```

Comment:
We'll want to keep these types parametrized; that probably means defining `class SampledFromStrategy(SearchStrategy[Ex]):`